libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt

```diff
@@ -1588,6 +1530,159 @@ internal class MapRouteLine(
+                        ?.flatten()
+                        ?.mapNotNull { it.intersections() }
+                        ?.flatten()
+                        ?.filter {
+                            it.geometryIndex() != null && it.mapboxStreetsV8()?.roadClass() != null
+                        }?.toList()?.forEach {
+                            roadClassMap[it.geometryIndex()!!] =
+                                it.mapboxStreetsV8()!!.roadClass()!!
+                        }
+
+                    var runningDistance = 0.0
+                    val routeLineTrafficData = mutableListOf<RouteLineTrafficExpressionData>()
+                    route.legs()?.mapNotNull {
+                        it.annotation()?.congestion()
+                    }?.flatten()?.forEachIndexed { index, congestion ->
+                        val roadclass = getRoadClassForIndex(roadClassMap, index)
+                        if (index == 0) {
+                            routeLineTrafficData.add(
+                                RouteLineTrafficExpressionData(
+                                    0.0,
+                                    congestion,
+                                    roadclass
+                                )
+                            )
+                        } else {
+                            runningDistance += distanceList[index - 1]
+                            if (routeLineTrafficData.last().trafficCongestionIdentifier ==
+                                congestion &&
+                                routeLineTrafficData.last().roadClass == roadclass
+                            ) {
+                                // continue
+                            } else if (routeLineTrafficData.last().trafficCongestionIdentifier ==
+                                congestion &&
+                                roadclass == null
+                            ) {
+                                // continue
+                            } else {
+                                routeLineTrafficData.add(
+                                    RouteLineTrafficExpressionData(
+                                        runningDistance,
+                                        congestion,
+                                        roadclass
+                                    )
+                                )
+                            }
+                        }
+                    }
+                    return routeLineTrafficData
+                }
+                Timber.i(
+                    "Route leg annotation does not contain distance and/or congestion " +
+                        "so route line traffic cannot be calculated."
+                )
+                return listOf()
+            }
+
+        private fun getRoadClassForIndex(roadClassMap: Map<Int, String>, index: Int): String? {
+            var indexCountdown = index
+            while(indexCountdown >= 0) {
+                if (roadClassMap.containsKey(indexCountdown)) {
```

Comment:
In order for this to be correct, you would need to add all intersection to this map, not only the ones that have road classes.